File: 2016/src/day01/task2/Task2.java
package day01.task2;

import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.time.Instant;
import java.util.HashSet;

public class Task2 {
   Path path = Path.of("inputs", "day01", "input.txt");
   Direction direction = Direction.North;
   HashSet<Vector2D> visited = new HashSet<>();
   Vector2D location = new Vector2D(0, 0);

   void task() {
      try {
         var input = Files.readString(this.path).trim().split(", ");
         for (var instruction : input) {
            var turnDirection = instruction.charAt(0);
            var distanceString = instruction.substring(1, instruction.length());
            var distance = Integer.parseInt(distanceString);

            this.direction = switch (turnDirection) {
               case 'R' -> this.direction.turnRight();
               case 'L' -> this.direction.turnLeft();
               default -> throw new RuntimeException("Invalid Direction");
            };
            if (this.moveAndCheckForDestination(distance)) {
               break;
            }
         }

         System.out.println(this.location.getDistanceFromOrigin());
      } catch (Exception e) {
         System.err.println(e);
         System.exit(1);
      }
   }

   boolean moveAndCheckForDestination(int distance) {
      for (int i = 0; i < distance; i++) {
         switch (this.direction) {
            case North -> this.location = new Vector2D(this.location.x(), this.location.y() - 1);
            case South -> this.location = new Vector2D(this.location.x(), this.location.y() + 1);
            case East -> this.location = new Vector2D(this.location.x() + 1, this.location.y());
            case West -> this.location = new Vector2D(this.location.x() - 1, this.location.y());
         }

         if (this.visited.contains(this.location)) {
            return true;
         }
         this.visited.add(this.location);
      }
      return false;
   }

   public static void main(String[] args) {
      var start = Instant.now();

      new Task2().task();

      System.out.println("Done in " + Duration.between(start, Instant.now()).toString().substring(2).toLowerCase());
   }
}

record Vector2D(int x, int y) {
   int getDistanceFromOrigin() {
      return Math.abs(this.x) + Math.abs(this.y);
   }
}

enum Direction {
   North,
   South,
   East,
   West;

   Direction turnRight() {
      return switch (this) {
         case North -> East;
         case South -> West;
         case East -> South;
         case West -> North;
      };
   }

   Direction turnLeft() {
      return switch (this) {
         case North -> West;
         case South -> East;
         case East -> North;
         case West -> South;
      };
   }
}

File: 2017/src/day11/task2.ts
import path from "node:path";

const PATH = path.join("inputs", "day11", "input.txt");

async function task2() {
   const input = (await Bun.file(PATH).text()).trim().split(",");
   const pos = { x: 0, y: 0 } satisfies Position;

   let result = 0;
   for (const direction of input) {
      switch (direction) {
         case "n":
            pos.x += 2;
            break;
         case "s":
            pos.x -= 2;
            break;
         case "ne":
            pos.x += 1;
            pos.y += 1;
            break;
         case "nw":
            pos.x += 1;
            pos.y -= 1;
            break;
         case "se":
            pos.x -= 1;
            pos.y += 1;
            break;
         case "sw":
            pos.x -= 1;
            pos.y -= 1;
            break;
      }

      const absPos = {
         x: Math.abs(pos.x),
         y: Math.abs(pos.y),
      } satisfies Position;

      let distance = Math.min(absPos.x, absPos.y);
      absPos.x -= distance;
      absPos.y -= distance;

      if (absPos.x) {
         distance += absPos.x / 2;
      }
      if (absPos.y) {
         distance += absPos.y;
      }

      result = Math.max(result, distance);
   }

   console.log(result);
}

type Position = {
   /** 1 per diagonal */
   x: number;
   /** 1 per diagonal */
   y: number;
};

console.time("Task Done");
await task2();
console.timeEnd("Task Done");
